src/Client/ConnectionString.cpp

```diff
@@ -0,0 +1,218 @@
+#include "ConnectionString.h"
+
+#include <Common/Exception.h>
+#include <Poco/Exception.h>
+#include <Poco/URI.h>
+
+#include <array>
+#include <iostream>
+#include <string>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+}
+
+namespace
+{
+
+using namespace std::string_literals;
+using namespace std::literals::string_view_literals;
+
+constexpr auto CONNECTION_URI_SCHEME = "clickhouse:"sv;
+
+void uriDecode(std::string & uri_encoded_string, bool plus_as_space)
+{
+    std::string temp;
+    Poco::URI::decode(uri_encoded_string, temp, plus_as_space);
+    std::swap(temp, uri_encoded_string);
+}
+
+void getHostAndPort(const Poco::URI & uri, std::vector<std::vector<std::string>> & hosts_and_ports_arguments)
+{
+    auto host = uri.getHost();
+    std::vector<std::string> host_and_port;
+    if (!host.empty())
+    {
+        uriDecode(host, false);
+        host_and_port.push_back("--host="s + host);
+    }
+
+    // Port can be written without host (":9000"). Empty host name equals to default host.
+    auto port = uri.getPort();
+    if (port != 0)
+        host_and_port.push_back("--port="s + std::to_string(port));
+
+    if (!host_and_port.empty())
+        hosts_and_ports_arguments.push_back(std::move(host_and_port));
+}
+
+void getHostAndPort(
+    Poco::URI & uri,
+    std::vector<std::vector<std::string>> & hosts_and_ports_arguments,
+    std::string_view host_and_port,
+    std::string_view right_part)
+{
+    // User info does not matter in sub URI
+    auto uri_string = std::string(CONNECTION_URI_SCHEME);
+    if (!host_and_port.empty())
```

Comment:
So `//` is only appended if a host/port was specified. This allows things like `clickhouse:user/database` (no `//`). Intended?

Reply:
Not just host port, `//` is written before `authority` (`user_info `@ `host_port`) see (https://datatracker.ietf.org/doc/html/rfc3986#section-3.2)

`clickhouse:user/database` is wrong URI. user name must be preceeded by '//' otherwise 'user' will be read as 'uri_path' (database). and '/database' cannot be parsed as query parameters - so parser will fail.

`clickhouse://user/database` is valid URI
`clickhouse:/database`  is valid URI
`clickhouse:///database` is valid URI

Reply:
Done